File: field_force/field_force/page/store_visit_report/store_visit_report.py
# Copyright (c) 2022, Invento Software Limited and contributors
# For license information, please see license.txt
import json

import frappe
from field_force.field_force.page.utils import *
from field_force.field_force.report.utils import *


@frappe.whitelist()
def execute(filters=None):
    columns = get_columns()
    filters = json.loads(filters)
    data = get_absolute_data(filters)
    # print(data)
    return data, columns

def get_absolute_data(filters, export=False):
    data = get_query_data(filters)
    site_directory = get_site_directory_path()
    data_dict = {}
    index = 1

    for store_visit in data:
        set_image_url(store_visit, site_directory)
        data_key = f"{store_visit.server_date}_{store_visit.customer}_{store_visit.sales_person}"
        store_visit.cheated = 'Yes' if store_visit.cheated else 'No'

        if data_key not in data_dict.keys():
            store_visit.date = frappe.format_value(store_visit.server_date, 'Date')
            data_dict[data_key] = store_visit

            # data_dict[data_key]['checkin_time_'] = store_visit.server_time
            checkin_time = get_time_in_12_hour_format(store_visit.server_time)

            data_dict[data_key]['checkin_name'] = store_visit.name
            data_dict[data_key]['checkin_image'] = store_visit.image
            data_dict[data_key]['checkout_image'] = '/files/default-image.png'

            if export:
                store_visit.user = store_visit.user_fullname or store_visit.user
                data_dict[data_key]['checkin_time'] = checkin_time
            else:
                set_link_to_doc(store_visit, 'sales_person', 'sales-person')
                set_link_to_doc(store_visit, 'customer', 'customer')
                data_dict[data_key]['checkin_time'] = get_doc_url('store-visit', store_visit.name, checkin_time)

            index += 1
        else:
            checkout_time = get_time_in_12_hour_format(store_visit.server_time)
            data_dict[data_key]['spent_time'] = get_spent_time(data_dict[data_key].server_time, store_visit.server_time)

            data_dict[data_key]['checkout_name'] = store_visit.name
            data_dict[data_key]['checkout_image'] = store_visit.image

            if export:
                data_dict[data_key]['checkout_time'] = checkout_time
            else:
                data_dict[data_key]['checkout_time'] = get_doc_url('store-visit', store_visit.name, checkout_time)

    if data_dict.values():
        return list(data_dict.values())[::-1]

    return []

def get_query_data(filters):
    conditions = get_conditions(filters)

    base_query = '''select store_visit.name, store_visit.user, store_visit.user_fullname,
                    store_visit.server_date, store_visit.server_time, store_visit.device_time,
                    store_visit.type, store_visit.image, store_visit.cheated, store_visit.customer,
                    store_visit.sales_person, store_visit.latitude, store_visit.longitude
                    from `tabStore Visit` store_visit where %s order by store_visit.server_date,
                    store_visit.server_time''' % conditions

    query_result = frappe.db.sql(base_query, as_dict=1, debug=0)
    return query_result


def get_conditions(filters):
    from_date = filters.get('from_date')
    to_date = filters.get('to_date')
    sales_person = filters.get('sales_person')
    customer = filters.get('customer')

    conditions = []

    if from_date:
        conditions.append('store_visit.server_date >= "%s"' % from_date)
    if to_date:
        conditions.append('store_visit.server_date <= "%s"' % to_date)
    if sales_person:
        conditions.append('store_visit.sales_person = "%s"' % sales_person)
    if customer:
        conditions.append('store_visit.customer = "%s"' % customer)

    return " and ".join(conditions)

def get_columns():
    columns =  [
        {'fieldname': 'sl', 'label': 'SL', 'width':30, 'expwidth': 5, 'export': False},
        {'fieldname': 'date', 'label': 'Date', 'width': 120, 'expwidth': 13},
        {'fieldname': 'sales_person', 'label': 'Sales Person', 'width': 220, 'expwidth': 25},
        {'fieldname': 'customer', 'label': 'Customer', 'width': 270, 'expwidth': 35},
        {'fieldname': 'checkin_time', 'label': 'IN Time', 'width': 120, 'expwidth': 15},
        {'fieldname': 'checkout_time', 'label': 'OUT Time', 'width': 120, 'expwidth': 15},
        {'fieldname': 'spent_time', 'label': 'Spent Time', 'width': 130, 'expwidth': 15},
        {'fieldname': 'cheated', 'label': 'Cheated', 'fieldtype': 'Data', 'width': 100, 'expwidth': 15},
        {'fieldname': 'checkin_image', 'label': 'IN Image', 'fieldtype': 'Image', 'width': 120, 'expwidth': 15, 'export': False},
        {'fieldname': 'checkout_image', 'label': 'OUT Image', 'fieldtype': 'Image', 'width': 120, 'expwidth': 15, 'export': False},
    ]
    return columns

@frappe.whitelist()
def export_file(**filters):
    columns = get_columns()
    data = get_export_data(filters)
    file_name = 'Store_Visit_Report.xlsx'
    generate_excel_and_download(columns, data, file_name, height=20)

def get_export_data(filters):
    query_result = get_absolute_data(filters, export=True)
    return query_result

File: field_force/field_force/page/store_visit_report/store_visit_report.js
frappe.pages['store-visit-report'].on_page_load = function(wrapper) {
  frappe.require("assets/field_force/css/page_modal.css");

    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Store Visit Report',
        single_column: true
    });

    window.addEventListener("click", function(event) {
      if (event.target.id !== "image") {
        document.getElementById('modal_section').style.display = 'none';
      }
    });

    $('.page-body').css('background', '#FFFFFF');
    new DailyAppUserAttendanceReport(page);
};

class DailyAppUserAttendanceReport {
    constructor(page) {
        this.page = page;
        this.make_form();
        this.make_menu();
        this.initialize_modal();
        // this.open_image_modal()
    }

    make_menu = () => {
        this.page.add_action_icon("refresh", () => {
            this.fetch_and_render();
        });
        this.page.add_menu_item("Export", ()=>{
            this.export_excel()
        })
    }
    make_form = () => {
        this.form = new frappe.ui.FieldGroup({
            fields: [
                {

                    fieldname: 'from_date',
                    label: __('From Date'),
                    fieldtype: 'Date',
                    default: frappe.datetime.get_today(),
                    change: () => this.fetch_and_render(),
                    reqd: 1,
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldname: 'to_date',
                    label: __('To Date'),
                    fieldtype: 'Date',
                    default: frappe.datetime.get_today(),
                    change: () => this.fetch_and_render(),
                },
                {
                    fieldtype: 'Column Break'
                },
                 {
                    fieldname: 'sales_person',
                    label: __('Sales Person'),
                    fieldtype: 'Link',
                    options: 'Sales Person',
                    change: () => this.fetch_and_render(),
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldname: 'customer',
                    label: __('Customer'),
                    fieldtype: 'Link',
                    options: 'Customer',
                    change: () => this.fetch_and_render(),
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldtype: 'Section Break'
                },
                {
                    fieldtype: 'HTML',
                    fieldname: 'preview'
                }
            ],
            body: this.page.body
        });
        this.form.make();
    }
    initialize_modal = () => {
      let html = `<div id="modal_section" class="modal_">
          <div class="modal-image">
          <img class="modal-content_" id="img" alt="img">
          </div>
        </div>`
        // this.form.body.append(html);
        $('.main-section').append(html)
    }
    fetch_and_render = () => {
        let {from_date, to_date, sales_person, customer} = this.form.get_values();
        if (!from_date) {
            this.form.get_field('preview').html('');
            return;
        }
        this.form.get_field('preview').html(`
        	<div class="text-muted margin-top">
        		${__("Fetching...")}
        	</div>
        `);
        frappe.call('field_force.field_force.page.store_visit_report.store_visit_report.execute', {
            filters: {
                from_date: from_date,
                to_date: to_date,
                sales_person: sales_person,
                customer: customer
            },
            freeze: true
        }).then(r => {
            let diff = r.message[0];
            let fields = r.message[1];
            this.render(diff, fields);
        });
    }

    render = (diff, fields) => {
        let table_header = this.table_header(fields);
        let table_body = this.table_body(diff, fields);
        this.form.get_field('preview').html(`<table class="table table-bordered" id="export_excel">${table_header}${table_body}</table>`);
    }
    table_header = (headers) => {
        let table_header = `<thead><tr>`;

        headers.forEach(function (data, index){
            table_header += `<th scope="col" width="${data.width}">${data.label || ''}</th>`;
        })

        table_header += `</tr>`;
        return table_header;
    }

    table_body = (diff, fields) => {
        var html = `<tbody>`;
        var serial_number = 1;

        diff.forEach(function (data, index) {
            html += `<tr>`;

            fields.forEach(function (field, index){
              if (field.fieldname === 'sl'){
                html += get_absolute_format_and_html(field, serial_number);
              }
              else {
                html += get_absolute_format_and_html(field, data[field.fieldname]);
              }
            })

            html += `</tr>`;
            serial_number = serial_number + 1;
        })

        html += `</tbody>`;
        return html
    }

    export_excel = () => {
        let {from_date, to_date, sales_person, customer} = this.form.get_values();
        let url = `/api/method/field_force.field_force.page.store_visit_report.store_visit_report.export_file`;
        url += `?from_date=${from_date||''}&to_date=${to_date||''}&sales_person=${sales_person||''}&customer=${customer||''}`;
        window.open(url, '_blank');
    }
}

function get_absolute_format_and_html(field, value){
    if (field.fieldtype === "Image"){
        return get_image_html(value);
    }
    else if (field.fieldtype === "Currency"){
        return get_currency_format(value);
    }
    else if (field.fieldtype ==="Data") {
        return `<td>${value || ''}</td>`;
    }
    else {
        return `<td>${value || ''}</td>`;
    }
}

function get_image_html(image_url) {
    let rotate = "transform: rotate(-90deg)"

    if (image_url === "/files/default-image.png"){
        rotate = '';
    }

    return `
        <td style="height:100px; width:120px;">
            <a href="#">
                <img style="height:100%; width:100%;" src="${image_url}" id="image" img-path="${image_url}" onclick="(
                    function(e){
                        document.getElementById(\'modal_section\').style.display=\'block\';
                        var nAgt = navigator.userAgent;
                        if (nAgt.indexOf('Safari') !== -1) {
                            document.getElementById(\'img\').src=e.target.currentSrc;
                        }else{
                            document.getElementById(\'img\').src=e.path[0].currentSrc;
                        }
                        return false;
                    }
                )(arguments[0]);
                return false;">
            </a>
        </td>`;
}

function get_currency_format(value){
    const company = frappe.defaults.get_user_default("company");
    const currency = frappe.get_doc(":Company", company).default_currency;
    let value_in_currency = format_currency(value, currency);
    return `<td>${value_in_currency || ''}</td>`
}
